contrib/epee/src/net_ssl.cpp

```diff
@@ -567,6 +569,49 @@ bool ssl_support_from_string(ssl_support_t &ssl, boost::string_ref s)
   return true;
 }
 
+boost::system::error_code store_ssl_keys(boost::asio::ssl::context& ssl, const boost::filesystem::path& base)
+{
+  EVP_PKEY* ssl_key = nullptr;
+  X509* ssl_cert = nullptr;
+  const auto ctx = ssl.native_handle();
+  if (!ctx || !(ssl_key = SSL_CTX_get0_privatekey(ctx)) || !(ssl_cert = SSL_CTX_get0_certificate(ctx)) || !base.has_filename())
```

Comment:
I don't understand. This returns an error, and that error is logged (no assert here)?

The `assert` above is to never fail, and represents a coding error. The `shared_ptr` is never `NULL`, and the `assert` is declaring intent for anyone making changes.

Reply:
> The assert above is to never fail, and represents a coding error. The shared_ptr is never NULL, and the assert is declaring intent for anyone making changes.

Typically `assert`s are used (by me and some standardish coding practices) only internally within a class, but not from external values (i.e. defensive programming). That class is large, so maybe the invariant goes unnoticed. There's a few getter's that would need updating too though, and it adds to the code generation for the failure case. Which stinks because everything is basically in the header, without some major refactor.

Reply:
I meant somehting like:

CHECK_AND_ASSERT_MES(ctx, {EINVAL, boost::system::system_category()}, "Context is NULL");

It just makes it easier to debug with logs, and it done pretty much everywhere in monero.

Reply:
Switched two of the checks to this macro.

These macros and log macros are frustrating when trying to use as a library for downstream components. You have to accept the logging utility as-is, or do a complete re-compile with different macro definitions. It would be nice to roll-back their usage in crypto and tx code to librarize components. The easiest to change might be the macros that throw; the `std::exception` just needs the line number and filename and it has all of the information contained in the log message.